File: problems/07_trees/0110_balanced_binary_tree/solution.dart
// Problem: 0110. Balanced Binary Tree
// Difficulty: Easy

class TreeNode {
  int val; TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root]; var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

int _depth(TreeNode? n) {
  if (n == null) return 0;
  final L = _depth(n.left);
  if (L == -1) return -1;
  final R = _depth(n.right);
  if (R == -1 || (L - R).abs() > 1) return -1;
  return 1 + (L > R ? L : R);
}


void main() {
  // Test 1
  // final result = solve(...);
  print('All tests passed!');
}

File: problems/07_trees/0110_balanced_binary_tree/solution.swift
// Problem: 0110. Balanced Binary Tree
// Difficulty: Easy

import Foundation

class TreeNode {
    var val: Int; var left: TreeNode?; var right: TreeNode?
    init(_ v: Int = 0, _ l: TreeNode? = nil, _ r: TreeNode? = nil) { val = v; left = l; right = r }
}

func fromLevel(_ a: [Int?]) -> TreeNode? {
    if a.isEmpty { return nil }
    let root = TreeNode(a[0]!); var q = [root]; var qi = 0, i = 1
    while qi < q.count && i < a.count {
        let n = q[qi]; qi += 1
        if i < a.count, let v = a[i] { n.left = TreeNode(v); q.append(n.left!) }
        i += 1
        if i < a.count, let v = a[i] { n.right = TreeNode(v); q.append(n.right!) }
        i += 1
    }
    return root
}

class Solution {
    func isBalanced(_ root: TreeNode?) -> Bool {
        func depth(_ n: TreeNode?) -> Int {
            guard let n = n else { return 0 }
            let L = depth(n.left)
            if L == -1 { return -1 }
            let R = depth(n.right)
            if R == -1 || abs(L - R) > 1 { return -1 }
            return 1 + max(L, R)
        }
        return depth(root) != -1
    }
}

func checkBool(_ a: Bool, _ e: Bool, _ msg: String) {
    if a != e { fatalError("FAIL \(msg)") }
}

let sol = Solution()
checkBool(sol.isBalanced(fromLevel([3, 9, 20, nil, nil, 15, 7])), true, "ex1")
checkBool(sol.isBalanced(fromLevel([1, 2, 2, 3, 3, nil, nil, 4, 4])), false, "ex2")
checkBool(sol.isBalanced(fromLevel([])), true, "empty")
checkBool(sol.isBalanced(fromLevel([1])), true, "one")
checkBool(sol.isBalanced(fromLevel([1, 2, nil, 3, nil, 4])), false, "skew")
print("All tests passed!")
